doc/mlsys_paper/paper.tex

```diff
@@ -301,48 +305,50 @@ \section{Types of objective functions}
+function, they only need to provide implementations of $f(x)$ and $f'(x)$, and
 then they can use any of the gradient-based optimizers that {\tt ensmallen}
-provides.  Table~\ref{tab:functionality} compares
-the classes of objective functions that {\tt ensmallen} can handle with other
-popular frameworks and libraries.
+provides.  Table~\ref{tab:functionality} contrasts 
+the classes of objective functions that can be handled by {\tt ensmallen}
+and other popular frameworks and libraries.
 
-Of course, not every optimization algorithm provided by {\tt ensmallen} can be
+Not every optimization algorithm provided by {\tt ensmallen} can be
 used by every class of objective function; for instance, a gradient-based
 optimizer such as L-BFGS cannot operate on a non-differentiable objective
 function.
-
-Therefore, the best the library can attain is to maximize the flexibility
+As such, the best the library can attain is to maximize the flexibility
 available, so that a user can easily implement a function $f(x)$ and have it
 work with as many optimizers as possible.
-To do this, {\tt ensmallen} makes heavy use of C++ template metaprogramming.
-When implementing an objective function to be
-optimized, a user can implement only a few methods and we can use
-metaprogramming to check that the given functions match the requirements of the
+
+To accomplish the flexibility, {\tt ensmallen} makes heavy use of
+C++ template metaprogramming.
+When implementing an objective function to be optimized,
+a user can implement only a few methods; metaprogramming is then automatically
+used to check that the given functions match the requirements of the
 optimizer that is being used.  When implementing an optimizer, we can assume
 that the given function to be optimized meets the required assumptions of the
-optimizers, and encode those requirements as \texttt{static\_assert} checks.
+optimizers, and encode those requirements as compile-time checks
+(via \texttt{static\_assert}).
 
-For the most common case of a differentiable $f(x)$, the user need only
+For the most common case of a differentiable $f(x)$, the user only needs to
 implement two methods:
 
 \vspace*{-0.3em}
 \begin{itemize} \itemsep -1pt
   \item \texttt{double Evaluate($x$)}: given coordinates $x$, this function
-should return the value of $f(x)$.
+returns the value of $f(x)$.
   \item \texttt{void Gradient($x$, $g$)}: given coordinates $x$ and a reference
 to $g$, set $g = f'(x)$.
 \end{itemize}
 \vspace*{-0.3em}
 
-Or, alternately, they can simply implement a \texttt{EvaluateWithGradient()}
+Alternatively, the user can simply implement a \texttt{EvaluateWithGradient()}
 function that computes both $f(x)$ and $f'(x)$ simultaneously, which is useful
 in cases where both the objective and gradient depend on similar computations.
 
-The required API for separable differentiable objective functions (i.e. those
+The required API for separable differentiable objective functions (ie., those
```

Comment:
Sorry for the pedantry :)  I think this should be "i.e.": https://english.stackexchange.com/questions/3911/correct-spelling-italicization-of-e-g-i-e

Reply:
i'm fine with either.  the only issue is the extra spacing that latex inserts after a full stop -- this is why i used a comma here.
how about using the ~ character to prevent the extra space. for example, instead of "i.e. those", use "i.e.~those".